File: Projects/Project-Euler/P14.py
target = 1000000
terms = {1:1}

def count_terms(n):
  if not terms.get(n):
    terms[n] = count_terms(3 * n + 1 if n % 2 else n // 2) + 1
  return terms.get(n)

max = 1

for i in range(2, target):
  if count_terms(i) > terms.get(max):
    max = i

print(max)

File: Projects/Project-Euler/P14.js
m => {
  let
  max = 1,
  terms = [, 1];
  
  const f = i => {
    let next;
    if (i % 2) {
      next = i * 3 + 1;
    }
    else {
      next = i / 2;
    }
    if (!terms[next]) {
      f(next);
    }
    terms[i] = terms[next] + 1;
    return terms[next] + 1;
  }
  for (let i = 2; i < m; i++) {
    if (f(i) > terms[max]) {
      max = i;
    }
  }
  return max;
}